### PR TITLE
cmd, debian: switch to snap{,d} in ubuntu-core, restart snapd on upgrade

### PR DESCRIPTION
**Pull request by chipaca**

### CODE REVIEW

**Comment by chipaca** [comment]
retest this please

**Comment by zyga** [comment]
Go parts look seductively easy. I cannot comment on the maintainer scripts, I'm not sure I know all the crazy edge cases that are possible there.

File: cmd/cmd.go [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+import (
+	"os"
+	"path/filepath"
+	"syscall"
+
+	"github.com/ubuntu-core/snappy/osutil"
+	"github.com/ubuntu-core/snappy/release"
+)
+
+const key = "SNAP_INNER"
```
Comment (niemeyer) [inline cmd/cmd.go]: `const reexecDisabledKey = "SNAP_REEXEC_DISABLED"`

This allows us to more pleasantly set from outside for testing as well, when necessary.

Comment (niemeyer) [inline cmd/cmd.go]: Or maybe just `SNAP_REEXEC`, and then compare to `0` so it defaults to `1`.

File: cmd/cmd.go [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+// SwitchToTheRealOne makes sure you're executing the "real"
+// binary. I.e. the one that ships in the ubuntu-core snap.
+func SwitchToTheRealOne() {
```
Comment (niemeyer) [inline cmd/cmd.go]: Can we please rename this to `ExecInCoreSnap`?  They are both very real, so it'll be easier to read if we drive some sort of distinction in the name.

File: cmd/cmd.go [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+		// you're already the real deal, natch
+		return
+	}
+
+	if os.Getenv(key) != "" {
+		os.Setenv(key, "")
+		return
+	}
+
+	exe, err := os.Readlink("/proc/self/exe")
```
Comment (niemeyer) [inline cmd/cmd.go]: Nice! I was thinking about using the argv, but this is better.

File: cmd/cmd.go [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+// SwitchToTheRealOne makes sure you're executing the "real"
+// binary. I.e. the one that ships in the ubuntu-core snap.
+func SwitchToTheRealOne() {
+	if !release.OnClassic {
+		// you're already the real deal, natch
+		return
+	}
+
+	if os.Getenv(key) != "" {
```
Comment (niemeyer) [inline cmd/cmd.go]: Can we please check for "1" specifically?  It's generally a bug to consider the existence of a variable at all to be indicative of a given choice, as it's an obvious thing to invert the logic (set it to `0`) and expect it to invert the semantic too.

File: debian/snapd.postinst [hunk 5]
```diff
@@ -0,0 +1,4 @@
+#!/bin/sh
+set -e
+systemctl daemon-reload
```
Comment (niemeyer) [inline debian/snapd.postinst]: These changes are nice and welcome in any case (assuming it works :-), but in a follow up we'll need to also do that when snapd detects a new ubuntu-core snap, which will be slightly interesting as it cannot just shoot its foot while acting.

**Comment by niemeyer** [comment]
That was quick, thank you. A few details, and LGTM.

File: cmd/cmd.go (hunk 4 above)
Comment (niemeyer) [inline cmd/cmd.go]: (fixed the test case)

File: debian/snapd.postinst (hunk 5 above)
Comment (chipaca) [inline debian/snapd.postinst]: should be fairly straightforward (but involved enough to be a separate branch): if in classic, tell the outer systemctl to restart snapd (with `--nowait`). If not in classic, do the reboot dance.

I think all of that is new code tho.

**Comment by mvo5** [comment]
Thanks for implementing this, this looks really cool! One thing we should ensure before landing this is that we should have a systemd job that updates all installed snaps automatically on classic. This ensures that we are not working with a stale ubuntu-core snap. The https://github.com/ubuntu-core/snappy/pull/1170 branch implements this, however it probably needs some more work before it can land.

**Comment by niemeyer** [comment]
retest this please

**Comment by niemeyer** [comment]
LGTM

**Comment by chipaca** [comment]
please oh please
retest this please
or don't, whatever, not like i care

File: cmd/snap/main.go [hunk 6]
```diff
@@ -157,6 +157,7 @@ func init() {
 }
 
 func main() {
+	cmd.ExecInCoreSnap()
```
Comment (mvo5) [inline cmd/snap/main.go]: Maybe `cmd.ReexecInCoreSnap()` ?

**Comment by mvo5** [comment]
Looks good!

**Comment by mvo5** [comment]
However we still need fully working `snap refresh` (i.e. 2.0.7) before this can be merged.

**Comment by niemeyer** [comment]
As discussed today in the meeting, we need to make the re-execution conditional on a particular version of the core snap, so we don't walk too far backwards when re-executing.

**Comment by niemeyer** [comment]
(probably in a separate branch?)

**Comment by fgimenez** [comment]
the images used for the tests were temporarily unavailable, they should be accessible now
retest this please

**Comment by fgimenez** [comment]
retest this please

**Comment by fgimenez** [comment]
retest this please

**Comment by fgimenez** [comment]
retest this please

**Comment by chipaca** [comment]
retest this please

**Comment by chipaca** [comment]
retest this please
but
gently

**Comment by zyga** [comment]
This is a historic moment folks :-)

**Pull request merged**